### PR TITLE
populate_local_conbench: simulate benchmark samples via semi-realistic distribution

### PR DESCRIPTION
**Pull request by jgehrcke**
Here: sample from normal distribution with aritificial lower bound left to the mean.
Add two different mechanisms for 'rare' instabilities.
This is an extremely rude approximation of reality.
This is not a PR I want to land, but this is just to show how I generated the data used for the screenshots in https://github.com/conbench/conbench/pull/591

### CODE REVIEW

**Comment by austin3dickey** [comment]
I'm not opposed to merging something like this. The better our dummy data in `populate_local_conbench.py`, the better we can make local changes. We should just make sure we're not losing anything by commenting out the other two `create_*()` lines.

File: conbench/tests/populate_local_conbench.py
Comment (jonkeane) [inline conbench/tests/populate_local_conbench.py:274]: ❤️ Oooph I've wanted this before! Thank you

File: conbench/tests/populate_local_conbench.py
Comment (jonkeane) [inline conbench/tests/populate_local_conbench.py:441]: I really love using simulated data. And I don't think we are _just yet_ using the shape of this data to test our techniques, so this comment is slightly less relevant just now but will be soon.

Could we make this generator function generate some data and save that as a fixture (we could also set a seed, I believe which would effectively be the same thing, we would get the same "random" numbers each time)? Even with pretty liberal z-scores, in principle there is a (very small!) chance that an observation is outside our band of variability when generated this way. Of course, we could re-try those tests and we would see that it's back to normal, but depending on what we are changing, the failure might start a wild goose chase! With a fixture (or seed), the tests outcomes themselves are deterministic, but we can re-shuffle the data if we want to or need to easily cause we have the generator helper

Comment (jgehrcke) [reply conbench/tests/populate_local_conbench.py:441]: Great point with respect to reproducibility in a test suite. Will certainly be of relevance. I agree that something like a regression detection method is best tested against specific data sets. Testing against certain distributions that become re-sampled upon every test run sounds pretty cool too, though (_if_ it leads to a stable outcome, it says a lot about the method in use).

**Comment by jgehrcke** [comment]
I have briefly cleaned this up, and we can now land this to have better history plot screenshots created by CI! Also motivated by https://github.com/conbench/conbench/pull/611#discussion_r1082866456.

This increases the time it takes to do `make db-populate` because of the GitHub API interaction. On my machine:
```
$ time make db-populate
...
real	2m29.229s
```

On HEAD of main:
```
$ time make db-populate
...
real	0m59.869s
```

At first I wanted to address this, but it's good to keep that work independent. This will be measurable win from improving architecture around fetching data about commits.

**Comment by jgehrcke** [comment]
I have reviewed the CI-generated history plot screenshot and it shows an error. This needs debugging and fixing before we can land this.

**Comment by jgehrcke** [comment]
Yeah, this seems to result in many GitHub API requests, hitting the quota in CI:
```
conbench-app-1  | [230126-16:43:46.393] [9] [conbench.entities.commit] INFO: got unexpected HTTP response with code 403: {"message":"API rate limit exceeded for installation ID 19416693.","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}
conbench-app-1  | [230126-16:43:46.393] [9] [conbench.entities.run] INFO: treat as unknown commit: error during get_github_commit(): 403 Client Error: Forbidden for url: https://api.github.com/repos/apache/arrow/commits/e26a88fd9fecca876073c5a18cf571c0daca6b3b
```

File: conbench/tests/populate_local_conbench.py
Comment (austin3dickey) [inline conbench/tests/populate_local_conbench.py]: nit: typo in `benchmars`

File: conbench/tests/populate_local_conbench.py
Comment (austin3dickey) [inline conbench/tests/populate_local_conbench.py:439]: Another slowdown idea: Could we slow down by a lot more, for a short period of time? To test the z-score code I'm working on, I've been using this synthetic data but essentially doubling the results for indexes 30-40. That's given me a great distribution-change to work with.

File: conbench/tests/populate_local_conbench.py
Comment (austin3dickey) [inline conbench/tests/populate_local_conbench.py]: Can you remove the empty strings out of this list? I think a few are showing up, because of how the string is formatted

File: conbench/tests/populate_local_conbench.py
Comment (austin3dickey) [inline conbench/tests/populate_local_conbench.py:453]: TIL the second argument to `enumerate()`!
>returns a tuple containing a count (from `start` which defaults to 0)

File: conbench/tests/populate_local_conbench.py
Comment (austin3dickey) [inline conbench/tests/populate_local_conbench.py]: Instead of every 20th commit, if you do consecutive commits it should hit the GitHub API a lot less often, because it won't need to fill in intermediate commits every time.

**Review by austin3dickey — APPROVED** [review]
Looking good!

File: conbench/tests/populate_local_conbench.py
Comment (jgehrcke) [reply conbench/tests/populate_local_conbench.py]: Thanks :)

But Mars is also cool! :D

File: conbench/tests/populate_local_conbench.py
Comment (jgehrcke) [reply conbench/tests/populate_local_conbench.py]: But I wanted to have a wider date range on the history plot! :-) 

For people reading along: when we depict an individual benchmark result on the history plot as of today then the position on the time axis is chosen by the _commit time_ of the commit corresponding to the benchmark run.

Related: https://github.com/conbench/conbench/issues/633

File: conbench/tests/populate_local_conbench.py
Comment (jgehrcke) [reply conbench/tests/populate_local_conbench.py]: will do!

File: conbench/tests/populate_local_conbench.py
Comment (jgehrcke) [reply conbench/tests/populate_local_conbench.py:439]: sure, that's a great starting point for simulating a distribution change (picking a certain time window that is transformed by a certain shift, instead of picking an individual sample)

File: conbench/tests/populate_local_conbench.py
Comment (austin3dickey) [reply conbench/tests/populate_local_conbench.py]: haha, that makes sense. Another option would be to choose a repo that has fewer commits over a longer time range. Maybe conbench itself?

**Comment by jgehrcke** [comment]
> Instead of every 20th commit, if you do consecutive commits it should hit the GitHub API a lot less often, because it won't need to fill in intermediate commits every time.

I did consecutive commits now towards actually landing this, and iterating later. I did not investigate too deeply, but the duration is about the same and I think the usage of the GitHub HTTP API is still in the same ballpark. Individual POST requests are responded to rather slowly. I think that is because there is a constant minimum number of GitHub HTTP API requests done for every posted result.

As of
```
default_branch = github.get_default_branch(...)
```
and also
```
commits = github.get_commits_to_branch(...)
```

```
conbench-app-1  | [230130-13:23:03.672] [42] [conbench.entities.commit] INFO: Finding all commits to the master branch of apache/arrow between  2023-01-03T09:51:12Z and 2023-01-03T12:23:30Z
conbench-app-1  | [230130-13:23:03.873] [42] [conbench.entities.commit] INFO: Backfilling 0 commit(s)
```

As we know there is a lot that can be architecturally optimized here for faster responses and fewer failure paths. But that work should not happen in this PR here.

I have added a change so that if Config.TESTING we fetch commits for the last 60 days only. Reduced the duration of `make db-populate` for me locally to 1 minute 40 s. 

Will merge this when CI passes.

**Comment by jgehrcke** [comment]
`make db-populate` was faster in CI, as of reduced latency to GitHub HTTP API:
```
Mon, 30 Jan 2023 13:34:25 GMT
230130-13:34:25.900 INFO: register via: http://127.0.0.1:32768/api/register/
Mon, 30 Jan 2023 13:34:26 GMT
230130-13:34:26.124 INFO: login via: http://127.0.0.1:32768/api/login/
Mon, 30 Jan 2023 13:34:26 GMT
230130-13:34:26.316 INFO: login succeeded

...
Mon, 30 Jan 2023 13:34:35 GMT
230130-13:34:35.454 INFO: Posted a benchmark with run_id 'history_machine12' and commit 88b42ef66fe664043c5ee5274b2982a3858b414e. Received status code 201. Took 1 attempt(s). Last attempt took 0.01044 s.
```

As of the few commits pushed in the past hour I did manage to consume CI's API quota again. Example from last CI run:
```
Mon, 30 Jan 2023 13:36:18 GMT
conbench-app-1  | [230130-13:34:35.405] [12] [conbench.entities.run] INFO: treat as unknown commit: error during get_github_commit(): 403 Client Error: Forbidden for url: https://api.github.com/repos/apache/arrow/commits/88b42ef66fe664043c5ee5274b2982a3858b414e
...
```

File: conbench/entities/commit.py
Comment (austin3dickey) [inline conbench/entities/commit.py:335]: Nice, thank you!

**Pull request merged**
